pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,142 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  // If the request timeout is reached on server side, the returned ReadResponse will contain all the datum responses that have been read.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
```

Comment:
```suggestion
  // Ack acknowledges a list of datum offsets.
  // When Ack is called, it implicitly indicates that the datum stream has been processed by the source vertex.
  // The caller (numa) expects the Ack to be successful, and it does not expect any errors. 
  // If there are some irrecoverable errors when the callee (UDSource) is processing the Ack,
  // then it is best to crash because there are no other retry mechanisms possible. 
  rpc Ack(AckRequest) returns (AckResponse);
```